scripts/sampinctrl.py

```diff
@@ -0,0 +1,228 @@
+
+    Args:
+        family: Atmel SAM family.
+        serie: Series.
+        variant: Variant information.
+
+    Returns:
+        Header file name.
+    """
+
+    sufix = ""
+    if revision:
+        sufix = f"X{revision}"
+
+    return f"sam{serie}{variant}{sufix}-pinctrl.h"
+
+
+def get_port_pin(pin_name):
+    """Obtain port and pin number from a pin name
+
+    Args:
+        pin_name: Pin name, e.g. PA0
+
+    Returns:
+        Port and pin, e.g. A, 0.
+    """
+
+    m = re.match(r"P([A-Z])(\d+)", pin_name.upper())
+    if not m:
+        raise ValueError(f"Unexpected pin name: {pin_name}")
+
+    return m.group(1), str(int(m.group(2)))
+
+
+def generate_atmel_sam_header(outdir, family, fmap, serie,
+                              variant, pin_cfgs, revision):
+    """Generate Atmel SAM header with pin configurations.
+
+    Args:
+        outdir: Output base directory.
+        family: Atmel SAM family.
+        fmap: Function to map pinctrl.
+        series: MCU Series.
+        variant: Variant information.
+        pin_cfgs: Pin configurations.
+    """
+
+    ofname = outdir / get_header_fname(serie, variant["pincode"], revision)
+    with open(ofname, "w") as f:
+        f.write(HEADER)
+        f.write(f'\n{"#include <dt-bindings/pinctrl/atmel_sam_pinctrl.h>"}\n')
+
+        if len(variant) > 2:
+            if variant["exception"]:
+                f.write(EXCEPTION)
+
+        for port, pin_num, mux, periph, signal in pin_cfgs:
+            f.write(f"\n/* p{port.lower()}{pin_num}{mux}_{periph}_{signal} "
+                    f"*/\n")
+            define = f"#define P{port.upper()}{pin_num.upper()}" \
+                     f"{mux.upper()}_{periph.upper()}_{signal.upper()}"
+            define_val = f"{fmap}('{port.upper()}', {pin_num}, " \
+                         f"'{mux.upper()}', {periph}, {signal})"
+            f.write(f"{define} \\\n\t{define_val}\n")
+
+
+def build_atmel_sam_sets(pin_cfgs, pin, pin_lst, serie, variant):
+    """Build Atmel SAM pin configurations sets.
+
+    Args:
+        serie: MCU Serie.
+        variant: Variant information.
+        pins: Pins description.
+
+    Returns:
+        Dictionary with pins configuration.
+    """
+
+    if len(pin_lst[0]) > 0:
+        for mux, periph, signal, *excludes in pin_lst:
+            if len(excludes) > 0:
+                if serie in excludes[0]:
+                    continue
+                if variant["pincode"] in excludes[0]:
+                    continue
+
+            port, pin_num = get_port_pin(pin)
+
+            pin_cfgs.append((port, pin_num, mux, periph, signal))
+
+
+def build_atmel_sam_pin_cfgs(serie, variant, pins):
+    """Build Atmel SAM pin configurations.
+
+    Args:
+        serie: MCU Serie.
+        variant: Variant information.
+        pins: Pins description.
+
+    Returns:
+        Dictionary with pins configuration.
+    """
+
+    pin_cfgs = []
+
+    pins = OrderedDict(natsorted(pins.items(), key=lambda kv: kv[0]))
```

Comment:
just curious: what does `natsorted` do?

Reply:
netsorted do true Human Order like p0a, p0b, p1a, p1b, p2a, p2b, p10b, p10d, p20a. So, doesn't matter how pinconfigs files will be defined a real human order is performed at Auto Generated files.